test/saithrift/tests/saitunnel.py

```diff
@@ -0,0 +1,321 @@
+    Tunnel Outer header dst  is 10.10.1.1
+    -----------------------------------------------------------------
+    Ingress side[port2]           |          Egress side[port1]
+    ------------------------------------------------------------------
+    ip's falls in 20.20.1.0       |        ip's falls in 10.10.1.0
+    ------------------------------------------------------------------
+    '''
+    def runTest(self):
+        print
+        switch_init(self.client)
+        time.sleep(10)
+        port1=port_list[0] #From where we expect encapsulated packet 
+        port2=port_list[1] #From where we send normal ip packet
+        mac1 = '00:05:06:00:00:00' # port1 l2 entry
+        mac2 = '00:05:06:00:00:02' 
+        #vr_mac='00:00:08:08:08:08'
+        mac_action=1                
+        vlan_id=0            #expecting to use unused vlans
+        addr_family=0 #1 for v6
+        tunnel_src_ip_addr='2.2.2.2' #src for outer ip
+        encap_ip_addr='10.10.1.1'    #egress enighbhor
+        ingress_ip_addr='20.20.1.1'  #ingress neighbor 
+        tunnel_ip_addr='1.1.1.0'     #to reach tunnel nhop
+        tunnel_ip_mask='255.255.255.0'
+        ingress_nhop_ip_addr='20.20.1.1'
+        egress_nhop_ip_addr='10.10.1.1'
+        nhop_ip_addr_ingress='20.20.1.0'
+        nhop_ip_mask_ingress='255.255.255.0'
+        nhop_ip_addr_encap='10.10.1.0'
+        nhop_ip_mask_encap='255.255.255.0'
+        initiator_ip_addr='10.10.1.1'  #tunnel dest ip(outer header)
+        neighbor_mac_ingress='00:05:06:00:00:02'
+        neighbor_mac_encap='00:05:06:00:00:00' #egress side
+        ip_addr_decap_src='10.10.1.1'
+        ip_addr_decap_dst='2.2.2.2'
+               
+        ########################################################################
+        #Creating Virtual router 
+        #######################################################################
+        vr_id=sai_thrift_create_virtual_router(self.client, 1, 1)
+            
+        #creating a underlay interface in loopback
+        underlay_if = sai_thrift_create_router_interface(self.client, is_port=1, port_id=0, vr_id=vr_id, vlan_id=vlan_id, v4_enabled=1, v6_enabled=1, mac='', is_lb=1)
+        #creating an overlay interface in loopback
+        overlay_if = sai_thrift_create_router_interface(self.client, is_port=1, port_id=0, vr_id=vr_id, vlan_id=vlan_id, v4_enabled=1, v6_enabled=1, mac='', is_lb=1)
+        #creating a tunnel
+        tunnel_id=sai_thrift_create_tunnel(self.client, addr_family=addr_family, ip_addr=tunnel_src_ip_addr, underlay_if=underlay_if, overlay_if=overlay_if)
+      
+        ##############################################################################
+        #  Egress configurations
+        #  create router interface,
+        #  create neighbor
+        #  create nhop
+        #  create route
+        #   
+        ##############################################################################
+       
+        #encap router interface

```

Comment:
Please reorganize the test in following section. It's very challenaging to understand the test. I am unable to understand the test unless you organize in the following way.

1. create neighbor, nexthop, and local subnet route
2. create underlay route that points to the nexthop
3. create tunnel, and tunnel next hop
4. create overlay route that points to the tunnel next hop